### PR TITLE
chain: add source code URL and schema field to oracle script

### LINKED ISSUE #1231: Add source code URL and schema field to oracle script
Let's just follow what wasmd does. Should be simple and help at least in short term.

### PR DESCRIPTION
**Pull request by tansawit**
Fixes: #1231

### CODE REVIEW

File: chain/x/oracle/types/msgs_test.go [hunk 1]
```diff
@@ -316,7 +316,7 @@ func TestMsgEditDataSourceGetSignBytes(t *testing.T) {
 func TestMsgCreateOracleScript(t *testing.T) {
 	owner := sdk.AccAddress([]byte("owner"))
 	sender := sdk.AccAddress([]byte("sender"))
-	msg := NewMsgCreateOracleScript(owner, "oracle_script_1", "description", []byte("code"), sender)
+	msg := NewMsgCreateOracleScript(owner, "oracle_script_1", "description", []byte("code"), sender, "schema", "sourceCodeURLl")
```
Comment (Benzbeeb) [inline chain/x/oracle/types/msgs_test.go]: please test msg.Schema and msg.SourceCodeURL

File: chain/app/genesis.go [hunk 2]
```diff
@@ -153,6 +153,8 @@ func GetDefaultDataSourcesAndOracleScripts(owner sdk.AccAddress) json.RawMessage
 			oracleScript.name,
 			oracleScript.description,
 			code,
+			"schema",
```
Comment (Benzbeeb) [inline chain/app/genesis.go]: You can add schema and code URL in oracleScripts (line 124)

File: chain/app/db_app.go [hunk 3]
```diff
@@ -131,6 +131,8 @@ func (app *dbBandApp) InitChain(req abci.RequestInitChain) abci.ResponseInitChai
 			time.Now(),
 			0,
 			nil,
+			"schema",
```
Comment (Benzbeeb) [inline chain/app/db_app.go]: oracleScript.Schema

File: chain/app/db_app.go [hunk 4]
```diff
@@ -131,6 +131,8 @@ func (app *dbBandApp) InitChain(req abci.RequestInitChain) abci.ResponseInitChai
...
+			"SourceCodeURL",
```
Comment (Benzbeeb) [inline chain/app/db_app.go]: ditto

File: chain/x/oracle/types/oracle_script.go [hunk 5]
```diff
@@ -6,20 +6,23 @@ import (
...
 type OracleScript struct {
-	Owner       sdk.AccAddress `json:"owner"`       // Address authorized to edit this script.
-	Name        string         `json:"name"`        // Short string explaining this oracle script.
-	Description string         `json:"description"` // Longer string explaining what this does.
-	Code        []byte         `json:"code"`        // Owasm bytecode to be run on-chain.
+	Owner         sdk.AccAddress `json:"owner"`       // Address authorized to edit this script.
+	Name          string         `json:"name"`        // Short string explaining this oracle script.
+	Description   string         `json:"description"` // Longer string explaining what this does.
+	Code          []byte         `json:"code"`        // Owasm bytecode to be run on-chain.
+	Schema        string         `json:"schema"`
```
Comment (taobun) [inline chain/x/oracle/types/oracle_script.go]: Can you add a comment about these filed?

File: chain/app/genesis.go (hunk 2 above)
Comment (taobun) [reply chain/app/genesis.go]: Example schema of "Crypto price script (Borsh version" schema should be `{"Input": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"symbol\\", \\"string\\"], [\\"multiplier\\", \\"u64\\"] ] }", "Output": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"px\\", \\"u64\\"] ] }`

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/x/oracle/types/msgs_test.go [hunk 6]
```diff
@@ -316,13 +316,15 @@ func TestMsgEditDataSourceGetSignBytes(t *testing.T) {
...
-	msg := NewMsgCreateOracleScript(owner, "oracle_script_1", "description", []byte("code"), sender)
+	msg := NewMsgCreateOracleScript(owner, "oracle_script_1", "description", []byte("code"), sender, "schema", `{"Input": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"symbol\\", \\"string\\"], [\\"multiplier\\", \\"u64\\"] ] }", "Output": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"px\\", \\"u64\\"] ] }`)
 	require.Equal(t, RouterKey, msg.Route())
 	require.Equal(t, "create_oracle_script", msg.Type())
 	require.Equal(t, owner, msg.Owner)
 	require.Equal(t, "oracle_script_1", msg.Name)
 	require.Equal(t, []byte("code"), msg.Code)
 	require.Equal(t, sender, msg.Sender)
+	require.Equal(t, "schema", msg.Schema)
+	require.Equal(t, `{"Input": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"symbol\\", \\"string\\"], [\\"multiplier\\", \\"u64\\"] ] }", "Output": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"px\\", \\"u64\\"] ] }`, msg.SourceCodeURL)
```
Comment (Benzbeeb) [inline chain/x/oracle/types/msgs_test.go]: It looks like schema 🗡️

File: chain/x/oracle/keeper/querier_test.go [hunk 7]
```diff
@@ -452,9 +454,11 @@ func TestQueryOracleScriptsByStartIdAndNumberOfOracleScripts(t *testing.T) {
 		name := "oracle_script_" + strconv.Itoa(i)
 		description := "description_" + strconv.Itoa(i)
 		code := []byte("code" + strconv.Itoa(i))
+		schema := "schema"
```
Comment (Benzbeeb) [inline chain/x/oracle/keeper/querier_test.go]: change to "schema" + strconv.Itoa(i)

File: chain/x/oracle/keeper/querier_test.go [hunk 8]
```diff
@@ -452,9 +454,11 @@ func TestQueryOracleScriptsByStartIdAndNumberOfOracleScripts(t *testing.T) {
...
+		sourceCodeURL := "sourceCodeURL"
```
Comment (Benzbeeb) [inline chain/x/oracle/keeper/querier_test.go]: ditto

File: chain/x/oracle/keeper/querier_test.go [hunk 9]
```diff
@@ -452,9 +454,11 @@ func TestQueryOracleScriptsByStartIdAndNumberOfOracleScripts(t *testing.T) {
...
 		eachOracleScript := types.NewOracleScriptQuerierInfo(types.OracleScriptID(i), owner, name, description, code)
 
-		_, err := keeper.AddOracleScript(ctx, eachOracleScript.Owner, eachOracleScript.Name, eachOracleScript.Description, eachOracleScript.Code)
+		_, err := keeper.AddOracleScript(ctx, eachOracleScript.Owner, eachOracleScript.Name, eachOracleScript.Description, eachOracleScript.Code, schema, sourceCodeURL)
```
Comment (Benzbeeb) [inline chain/x/oracle/keeper/querier_test.go]: eachOracleScript.Schema and eachOracleScript.sourceCodeURL too

File: chain/cmd/request/main.go [hunk 10]
```diff
@@ -224,7 +224,7 @@ func main() {
 
 			fmt.Println(tx.SendTransaction(
 				oracle.NewMsgCreateOracleScript(
-					tx.Sender(), "Silly script", "Test oracle script", bytes, tx.Sender()),
+					tx.Sender(), "Silly script", "Test oracle script", bytes, tx.Sender(), "schema", "sourceCodeURL"),
```
Comment (Benzbeeb) [inline chain/cmd/request/main.go]: I think if we don't have schema and source code URL we can set it to an empty string 
@taobun any suggestion

File: chain/cmd/request/main.go [hunk 11]
```diff
@@ -97,7 +97,7 @@ func main() {
 				oracle.NewMsgCreateOracleScript(
 					tx.Sender(), "Crypto price script",
 					"Oracle script for getting an average crypto price from many sources.",
-					oracleBytes, tx.Sender(),
+					oracleBytes, tx.Sender(), "schema", "sourceCodeURL",
```
Comment (Benzbeeb) [inline chain/cmd/request/main.go]: `{"Input": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"symbol\\", \\"string\\"], [\\"multiplier\\", \\"u64\\"] ] }", "Output": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"px\\", \\"u64\\"] ] }`

**Review by Benzbeeb — CHANGES_REQUESTED** [review]

File: chain/app/genesis.go [hunk 12]
```diff
@@ -122,24 +122,32 @@ func GetDefaultDataSourcesAndOracleScripts(owner sdk.AccAddress) json.RawMessage
...
-		description string
-		path        string
+		name          string
+		description   string
+		path          string
+		schema        string
+		sourceCodeURL string
 	}{
 		{
 			"Crypto price script",
```
Comment (taobun) [inline chain/app/genesis.go]: Remove this script

File: chain/app/genesis.go [hunk 13]
```diff
@@ -122,24 +122,32 @@ func GetDefaultDataSourcesAndOracleScripts(owner sdk.AccAddress) json.RawMessage
...
 			"Oracle script for getting an average crypto price from many sources encoding parameter by borsh.",
 			"./owasm/res/crypto_price_borsh.wasm",
+			`{"Input": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"symbol\\", \\"string\\"], [\\"multiplier\\", \\"u64\\"] ] }", "Output": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"px\\", \\"u64\\"] ] }`,
+			`https://bandprotocol.com`,
 		},
 		{
 			"Gold price script",
 			"Oracle script for getting an average gold price in ATOM",
 			"./owasm/res/gold_price.wasm",
+			`{"Input": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"symbol\\", \\"string\\"], [\\"multiplier\\", \\"u64\\"] ] }", "Output": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"px\\", \\"u64\\"] ] }`,
```
Comment (taobun) [inline chain/app/genesis.go]: {"Input": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"multiplier\\", \\"u64\\"] ] }", "Output": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"px\\", \\"u64\\"] ] }

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/x/oracle/types/msgs.go [hunk 14]
```diff
@@ -289,11 +289,13 @@ func (msg MsgEditDataSource) GetSignBytes() []byte {
...
-	Owner       sdk.AccAddress `json:"owner"`
-	Name        string         `json:"name"`
-	Description string         `json:"description"`
-	Code        []byte         `json:"code"`
-	Sender      sdk.AccAddress `json:"sender"`
+	Owner         sdk.AccAddress `json:"owner"`
+	Name          string         `json:"name"`
+	Description   string         `json:"description"`
+	Code          []byte         `json:"code"`
+	Sender        sdk.AccAddress `json:"sender"`
```
Comment (sorawit) [inline chain/x/oracle/types/msgs.go]: Sender should be the last parameter.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/types/querier.go [hunk 15]
```diff
@@ -64,6 +64,8 @@ type OracleScriptQuerierInfo struct {
 	Name        string         `json:"name"`
 	Description string         `json:"description"`
 	Code        []byte         `json:"code"`
+	Schema     string `json:"schema"`
```
Comment (Benzbeeb) [inline chain/x/oracle/types/querier.go]: auto format, please

**Review by Benzbeeb — CHANGES_REQUESTED** [review]

File: chain/x/oracle/types/msgs_test.go [hunk 16]
```diff
@@ -316,13 +316,15 @@ func TestMsgEditDataSourceGetSignBytes(t *testing.T) {
 func TestMsgCreateOracleScript(t *testing.T) {
 	owner := sdk.AccAddress([]byte("owner"))
 	sender := sdk.AccAddress([]byte("sender"))
-	msg := NewMsgCreateOracleScript(owner, "oracle_script_1", "description", []byte("code"), sender)
+	msg := NewMsgCreateOracleScript(owner, "oracle_script_1", "description", []byte("code"), sender, `{"Input": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"symbol\\", \\"string\\"], [\\"multiplier\\", \\"u64\\"] ] }", "Output": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"px\\", \\"u64\\"] ] }`, `https://bandprotocol.com`)
```
Comment (taobun) [inline chain/x/oracle/types/msgs_test.go]: Sender to the last arguments

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by sorawit — DISMISSED** [review]

File: chain/x/oracle/keeper/oracle_script_test.go [hunk 17]
```diff
@@ -129,12 +132,16 @@ func TestAddOracleScriptTooLongDescription(t *testing.T) {
 	k.SetParam(ctx, types.KeyMaxDescriptionLength, 41)
 	_, err := k.AddOracleScript(ctx,
 		Owner.Address, BasicName, "________THIS_STRING_HAS_SIZE_OF_42________", BasicCode,
+		`{"Input": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"symbol\\", \\"string\\"], [\\"multiplier\\", \\"u64\\"] ] }", "Output": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"px\\", \\"u64\\"] ] }`,
+		"https://bandprotocol.com",
```
Comment (Benzbeeb) [inline chain/x/oracle/keeper/oracle_script_test.go]: BasicSchema,BasicSourceURL

File: chain/x/oracle/keeper/oracle_script_test.go [hunk 18]
```diff
@@ -129,12 +132,16 @@ func TestAddOracleScriptTooLongDescription(t *testing.T) {
...
 	)
 	require.Error(t, err)
 	// Sets max desc length to 42. We should now be able to add the oracle script.
 	k.SetParam(ctx, types.KeyMaxDescriptionLength, 42)
 	_, err = k.AddOracleScript(ctx,
 		Owner.Address, BasicName, "________THIS_STRING_HAS_SIZE_OF_42________", BasicCode,
+		`{"Input": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"symbol\\", \\"string\\"], [\\"multiplier\\", \\"u64\\"] ] }", "Output": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"px\\", \\"u64\\"] ] }`,
+		"https://bandprotocol.com",
```
Comment (Benzbeeb) [inline chain/x/oracle/keeper/oracle_script_test.go]: ditto

File: chain/x/oracle/keeper/oracle_script_test.go [hunk 19]
```diff
@@ -166,13 +177,17 @@ func TestAddOracleScriptTooBigCode(t *testing.T) {
 	_, err := k.AddOracleScript(ctx,
 		Owner.Address, BasicName, BasicDesc,
 		[]byte("________THIS_STRING_HAS_SIZE_OF_42________"),
+		`{"Input": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"symbol\\", \\"string\\"], [\\"multiplier\\", \\"u64\\"] ] }", "Output": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"px\\", \\"u64\\"] ] }`,
+		"https://bandprotocol.com",
```
Comment (Benzbeeb) [inline chain/x/oracle/keeper/oracle_script_test.go]: ditto

File: chain/x/oracle/keeper/oracle_script_test.go [hunk 20]
```diff
@@ -166,13 +177,17 @@ func TestAddOracleScriptTooBigCode(t *testing.T) {
...
 	)
 	require.Error(t, err)
 	// Sets max code size to 50. We should now be able to add the oracle script.
 	k.SetParam(ctx, types.KeyMaxOracleScriptCodeSize, 50)
 	_, err = k.AddOracleScript(ctx,
 		Owner.Address, BasicName, BasicDesc,
 		[]byte("________THIS_STRING_HAS_SIZE_OF_42________"),
+		`{"Input": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"symbol\\", \\"string\\"], [\\"multiplier\\", \\"u64\\"] ] }", "Output": "{ \\"kind\\": \\"struct\\", \\"fields\\": [ [\\"px\\", \\"u64\\"] ] }`,
+		"https://bandprotocol.com",
```
Comment (Benzbeeb) [inline chain/x/oracle/keeper/oracle_script_test.go]: ditto

**Review by Benzbeeb — CHANGES_REQUESTED** [review]

**Review by Benzbeeb — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**